creating pydra paper (#1)
nipype/scipy_proceedings: papers/pydra/paper.rst

that is presented in this paper.

The package is a part of the second generation of the *Nipype* ecosystem [TODO: ref]
--- an open-source framework that provides a uniform interface to existing neuroimaging
software and facilitates interaction between them.
Nipype project was born in the neuroimaging community, and has been helping scientists build
workflows for a decade, providing a uniform interface to such neuroimaging packages
as FSL [ref], ANTs [ref], AFNI [ref], FreeSurfer [ref] and SPM [ref].
This flexibility has made it an ideal basis for popular preprocessing tools,
such as fMRIPrep [ref] and C-PAC[ref].
The second generation of *Nipype* ecosystem is meant to provides additional flexibility
and is being developed with reproducibility, ease of use, and scalability in mind.
Pydra is a standalone project and is designed as a general-purpose dataflow engine
to support any scientific domain.

The goal of Pydra is to provide a lightweight dataflow engine for computational graph construction,
manipulation, and distributed execution, as well as ensuring reproducibility of scientific pipelines.

> In Pydra, a dataflow is represented as a graph, where each node is a Task object.

Reviewer: ```suggestion
In Pydra, a dataflow is represented as a directed acyclic graph, where each node represents a 
Python function, execution of an external tool, or another reusable dataflow.
```

The combination of several key features makes Pydra a customizable and powerful dataflow engine:

* Composable dataflows: any node of a dataflow graph can be another dataflow,

>   allowing for nested dataflows to arbitrary depths.

Reviewer: ```suggestion
  allowing for nested dataflows to arbitrary depths and encourages creating reusable dataflows.
```

* Flexible semantics for creating nested loops over input.
  Any Task or dataflow can be run over input parameter sets and the outputs can be recombined

>   (similar concept to Map-Reduce model, but applies to nested graphs).

Reviewer: ```suggestion
  (similar concept to Map-Reduce model, but Pydra extends this to graphs with nested dataflows).
```

* A content-addressable global cache: hash values are computed for each graph and each Task.
  Supports reuse of previously computed and stored dataflows and Tasks.

* Support for Python functions and external (shell) commands.

* Native container execution support. Any dataflow or Task can be executed in an associated
  container (via Docker or Singularity) enabling greater consistency for reproducibility.

> Pydra has a small set of dependencies, it is a pure Python package that depends mostly
> on the Python Standard Library, with an exception for *attr* package and testing framework,
> that is based on *pytest*.
> It is important to note, that Pydra is not designed to replace a pure Python to run
> single algorithms or functions.
> It is not design to help with algorithms that has to handle everything in the memory.
> Pydra is design to support file based computation instead.

Reviewer: ```suggestion
Pydra is a pure Python package with a limited set of dependencies, which are themselves only dependent on
the Python Standard library. Pydra uses the *attr* package for type annotation and validation of inputs and 
outputs of tasks, the *cloudpickle* package to pickle interactive task definitions, and the *pytest* testing 
framework. Pydra is intended to help scientific workflows which rely on significant file-based operations and 
which evaluate outcomes of complex dataflows over a hyper-space of parameters. It is important to note, that
Pydra is not a framework for writing efficient scientific algorithms or for use in applications where caching and 
distributed execution are not necessary. Since Pydra relies on a filesystem cache at present it is also not
designed for dataflows that need to operate purely in memory. 
```

The next section will describe the Pydra architecture --- main package classes

> and interaction between them. In *Key Feature* section focuses on several features
> that makes *Pydra* powerful and worth trying package.
> In the *Application* section, a few example will be presented ...[todo]

Reviewer: ```suggestion
and interactions between them. The *Key Features* section focuses on a set of features whose
combination distinguishes Pydra from other dataflow engines. The paper concludes with a set
of applied examples demonstrating the power and utility of Pydra.
```

At this moment *Pydra* supports three types of software: *ConcurrentFutures* [ref],
*Slurm* [ref] and *Dask* [ref].
Currently ``ConcurrentFuturesWorker`` has the biggest support, but ``SlurmWorker``
and ``DaskWorker`` are planned to have a full support.

Key Features
------------

In this section, chosen features of *Pydra* will be presented.
Some of the features are present in other packages, but the combination
of the following features makes *Pydra* a powerful tool in scientific computation.

Nested Workflows
================

*Workflows* in *Pydra* can contain multiple *Tasks*, but they are still *Tasks*,
and have all of the *Tasks* attributes and methods.
As a consequence, a *Workflow* can be also used as a node in the executable graph.
This provides an easy way of creating nested workflows of arbitrary depth,
as shown in Fig. :ref:`nested`.

.. figure:: nested_workflow-crop.pdf
   :figclass: h!
   :scale: 40%

   A nested Pydra workflow, black circles represent single Task,
   and Workflows are represented by red rectangles. :label:`nested`

State and Nested Loops over Input
=================================

> One of the main goal of *Pydra* wa sto support flexible creation

Reviewer 2: ```suggestion
One of the main goal of *Pydra* is to support flexible creation
```

It is also possible to combine all values of *x* for a specifc values of *y*, but it is also possible
to combine all the values by providing a list of fields to the combiner.
When a full combiner is set, i.e. all fields are in the combiner, the output is a one dimensional list:

.. math::
   :type: eqnarray

   \textcolor{red}{\mathnormal{C} = [x, y]} : out(x_1, y1), ...out(x_n, y_m)
    \longmapsto [out(x_1, y_1), ..., out(x_n, y_m)]

And is represented in Fig. :ref:`ndspl3comb3` (todo: should probably change a,b to x,y)

.. figure:: nd_spl_3_comb3-crop.pdf
   :figclass: h!
   :scale: 75%

   Diagram representing a Task with two input fields, an outer splitter and a full combiner. :label:`ndspl3comb3`

These are the basic examples of *splitters* and *combiners*, but *Pydra* allows for mixing
*splitters* and *combiners* on various level.
They could be set on a single *Task* level, or on *Workflow* level.
They could be also passed from one *Task* to the followings within a *Workflow*.

Global Cache
============

Applications and Examples
-------------------------

Machine Learning: Model Comparison
==================================

.. code-block:: python

>   ml example TODO

Reviewer 2: Just use the ml example from the [pycon poster](https://docs.google.com/presentation/d/10tS2I34rS0G9qz6v29qVd77OUydjP_FdBklrgAGmYSw/edit?usp=sharing)?    I'll add the code below and include an example workflow that uses a public dataset

> 

Reviewer 2: ```python
import pydra
import typing as ty
import os

# 1 - CREATE PYDRA TASKS
@pydra.mark.task # Convert python function to a Pydra Task object
@pydra.mark.annotate({"return": {"X": ty.Any, "Y": ty.Any, "groups": ty.Any}})  # Annotate to create named outputs
def read_file(filename, target_inds, x_indices=None, group='groups'):
   # Read and transform a CSV file to sklearn inputs
   import pandas as pd
   data = pd.read_csv(filename)
   X = data.iloc[:, x_indices]
   Y = data[target_inds]
   if group in data.keys():
       groups = data[:, [group]]
   else:
       groups = list(range(X.shape[0]))
   return X.values, Y.values, groups

@pydra.mark.task  # Convert python function to a Pydra Task object
@pydra.mark.annotate({"return": {"splits": ty.Any, "split_indices": ty.Any}})  # Annotate to create named outputs
def gen_splits(n_splits, test_size, X, Y, groups=None, random_state=0):
   # Generate a set of train-test splits
   from sklearn.model_selection import GroupShuffleSplit
   gss = GroupShuffleSplit(n_splits=n_splits, test_size=test_size,
                           random_state=random_state)
   train_test_splits = list(gss.split(X, Y, groups=groups))
   split_indices = list(range(n_splits))
   return train_test_splits, split_indices

@pydra.mark.task
@pydra.mark.annotate({"return": {"auc": ty.Any}})
def train_test_kernel(X, y, train_test_split, split_index, clf_info, permute):
   # Train and test a classifier on actual or permuted labels
   from sklearn.preprocessing import StandardScaler
   from sklearn.pipeline import Pipeline
   from sklearn.metrics import roc_auc_score
   mod = __import__(clf_info[0], fromlist=[clf_info[1]])
   clf = getattr(mod, clf_info[1])(**clf_info[2])
   if len(clf_info) > 3: # Run a GridSearch when param_grid available
       from sklearn.model_selection import GridSearchCV
       clf = GridSearchCV(clf, param_grid=clf_info[3])
   train_index, test_index = train_test_split[split_index]
   pipe = Pipeline([('std', StandardScaler()), (clf_info[1], clf)])
   y = y.ravel()
   if permute: # Run a generic permutation to create a null model
       pipe.fit(X[train_index], y[np.random.permutation(train_index)])
   else:
       pipe.fit(X[train_index], y[train_index])
   auc = roc_auc_score(y[test_index], pipe.predict(X[test_index]))
   return auc

# 2 - Specify inputs

# Suppose 'iris.csv' is the result of
from sklearn import datasets
X, y = datasets.load_iris(return_X_y=True)
iris_data = np.hstack((X, y.reshape((-1,1))))
print(iris_data.sample(10))

clfs = [
 ('sklearn.ensemble', 'ExtraTreesClassifier',
        dict(n_estimators=100, class_weight='balanced')),
 ('sklearn.neural_network', 'MLPClassifier', 
        dict(alpha=1, max_iter=1000)),
 ('sklearn.neighbors', 'KNeighborsClassifier', dict(),
        [{'n_neighbors': [3, 5, 7, 9, 11, 13, 15, 17, 19],
          'weights': ['uniform','distance']}]),
 ('sklearn.tree', 'DecisionTreeClassifier', 
        dict(max_depth=5)),
 ('sklearn.ensemble', 'AdaBoostClassifier', dict())]
inputs = {"filename": os.path.abspath('breast_cancer.csv'),
         "x_indices": range(30), "target_inds": -1,
         "n_splits": 5, "test_size": 0.2,
         "clf_info": clfs,  "permute": [True, False]
         }
n_procs = 8 # for parallel processing
cache_dir = os.path.join(os.getcwd(), 'cache')
wf_cache_dir = os.path.join(os.getcwd(), 'cache-wf')

# 3 - Use Pydra in Workflow style:

# Encapsulate tasks in a Workflow reuse script output cache
wf = pydra.Workflow(name="ml_wf", input_spec=list(inputs.keys()),
               **inputs, cache_dir=wf_cache_dir, # workflow cache 
                   cache_locations=[cache_dir]) # reuses script cache

# joint map over classifiers and permutation
wf.split(['clf_info', 'permute']) 
wf.add(read_file(name="readcsv",               # add task
                filename=wf.lzin.filename,     # connect workflow input
                x_indices=wf.lzin.x_indices,
                target_inds=wf.lzin.target_inds))
wf.add(gen_splits(name="gensplit",             # add task 
                 n_splits=wf.lzin.n_splits,    # connect workflow input
                 test_size=wf.lzin.test_size,
                 # connect lazy-eval outputs of previous task
                 X=wf.readcsv.lzout.X, Y=wf.readcsv.lzout.Y,
                 groups=wf.readcsv.lzout.groups))
wf.add(train_test_kernel(name="fit_clf", # use outputs from both tasks
                   X=wf.readcsv.lzout.X, y=wf.readcsv.lzout.Y,
                   train_test_split=wf.gensplit.lzout.splits,
                   split_index=wf.gensplit.lzout.split_indices,
                   clf_info=wf.lzin.clf_info, permute=wf.lzin.permute))
wf.fit_clf.split('split_index').combine('split_index') # Parallel spec
wf.set_output([("auc", wf.fit_clf.lzout.auc)]) # connect workflow output

#Execute the workflow in parallel using multiple processes
with pydra.Submitter(plugin="cf", n_procs=n_procs) as sub:
   sub(runnable=wf)

print(wf.result(return_inputs=True))

```

Author: it would be great if you can work on this part, however if you work on big snippet of the code, you should work on this locally and build the pdf paper.  You will see that unfortunately the lines of code have to be short to fit nicely to the column.

Author: so this is why i do not merge it here

:email: djarecka@gmail.com
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Mathias Goncalves
:email: todo
:institution: Stanford University, Stanford, CA, USA
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Christopher J. Markiewicz
:email:  todo
:institution: Stanford University, Stanford, CA, USA

:author: Oscar Esteban
:email: todo
:institution: Stanford University, Stanford, CA, USA

:author: Nicole Lo
:email: todo
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Jakub Kaczmarzyk
:email: todo
:institution: Stony Brook University School of Medicine, Stony Brook, NY, USA
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Satrajit Ghosh
:email: satra@mit.edu
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

--------------------------------------------------------------------------
Pydra - a flexible and lightweight dataflow engine for scientific analyses
--------------------------------------------------------------------------

.. class:: abstract

This paper presents a new lightweight dataflow engine written

> in Python: Pydra. TODO

Reviewer: ```suggestion
in Python: Pydra. TODO - check PR build
```

a different application:

* ``FunctionTask`` is a *Task* that executes Python functions. Most Python functions
  declared in an existing library, package, or interactively can be converted to
  a ``FunctionTask`` by using Pydra decorator - ``mark.task``.

  .. code-block:: python

     import numpy as np
     from pydra import mark
     fft = mark.annotate({'a': np.ndarray,
                      'return': float})(np.fft.fft)
     fft_task = mark.task(fft)()
     result = fft_task(a=np.random.rand(512))

  `fft_task` is now a Pydra task and result will contain Pydra ``Result`` object.
  In addition, the user can use Python's function annotation or another Pydra
  decorator |---| ``mark.annotate`` in order to specify the output. In the
  following example, we decorate an arbitrary Python function to create named
  outputs.

  .. code-block:: python

     @mark.task
     @mark.annotate(
         {"return": {"mean": float, "std": float}}
     )
     def mean_dev(my_data):
         import statistics as st
         return st.mean(my_data), st.stdev(my_data)

     result = mean_dev(my_data=[...])()

  When the task is executed `result.output` will contain two attributes: `mean`

>   and `std`. This naming allows references to pass different outputs to

Reviewer: ```suggestion
  and `std`. These named outputs allowing passing different outputs to
```

:author: Dorota Jarecka
:email: djarecka@gmail.com
:institution: Massachusetts Institute of Technology, Cambridge, MA, USA

:author: Mathias Goncalves

> :email: todo

Reviewer 3: ```suggestion
:email: mathiasg@stanford.edu
```

Introduction
------------

Scientific workflows often require sophisticated analyses that encompass
a large collection of algorithms.
The algorithms, that were originally not necessarily designed to work together,
often written by different authors in different times.
Some may be written in Python, while others might require calling external programs.
It is a common practice to create semi-manual workflows that require scientist
interaction to handle files and partial results between algorithms and external tools.
This approach is conceptually simple and easy to implement, but the resulting workflow
is often time consuming, error-prone and difficult to share with others.
Consistency, reproducibility and scalability demand scientific workflows be organized
into fully automated pipelines.
This was the motivation behind writting a new dataflow engine written in Python, *Pydra*,
that is presented in this paper.

The package is a part of the second generation of the *Nipype* ecosystem [TODO: ref]
--- an open-source framework that provides a uniform interface to existing neuroimaging
software and facilitates interaction between them.

> Nipype project was born in the neuroimaging community, and has been helping scientists build

Reviewer 3: ```suggestion
The *Nipype* project was born in the neuroimaging community, and has been helping scientists build
```
